### PR TITLE
fix(message): correctly set the done state in sym_encrypted reader

### PR DESCRIPTION
**Pull request by dignifiedquire**

### CODE REVIEW

**Comment by hko-s** [comment]
I've just run the fuzz test `1_message_from_bytes` on this branch (this PR was triggered by that test, in https://github.com/rpgp/rpgp/issues/542).

In the first 10 CPU minutes,  the fuzzer didn't find any complaints, but then it stopped with an OOM:

```
Failing input:

	artifacts/1_message_from_bytes/oom-02d73ac755c5679a709b393a2ac803d23a0bf0fc

Output of `std::fmt::Debug`:

	[141, 0, 137, 4, 11, 4, 191, 163, 4, 139, 245, 244, 110, 254, 136, 160, 4, 2, 160, 163, 1, 6, 14, 6, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 22, 22, 34, 22, 22, 163, 163, 1, 91, 163, 237, 0, 249, 0, 2, 0, 81, 58, 36, 0, 36, 163, 29, 0, 22, 22, 0, 22, 64, 147, 0, 167, 89]
```

I'll add a unit test for it.

**Comment by hko-s** [comment]
> In the first 10 CPU minutes, the fuzzer didn't find any complaints, but then it stopped with an OOM:

False alarm. The fuzzer made an SKESK using Argon2, which asks for 2 gbyte of RAM, and that apparently exceeds the memory limit that libFuzzer sets:

```
Running: artifacts/1_message_from_bytes/oom-02d73ac755c5679a709b393a2ac803d23a0bf0fc
==1016183== ERROR: libFuzzer: out-of-memory (used: 2125Mb; limit: 2048Mb)
   To change the out-of-memory limit use -rss_limit_mb=<N>
```

So as far as we can tell, this PR fixes the previously existing fuzzer issue.

**Comment by hko-s** [comment]
> So as far as we can tell, this PR fixes the previously existing fuzzer issue.

Mostly as a note to self: I've now run ~70 CPU minutes of `cargo +nightly fuzz run 1_message_from_bytes -- -rss_limit_mb=8200` without incident.

File: src/composed/message/reader/sym_encrypted.rs [hunk 1]
```diff
@@ -93,14 +93,56 @@ impl<R: BufRead> SymEncryptedDataReader<R> {
...
+            }
+        }
+
+        Ok(())
+    }
 }
 
 impl<R: BufRead> BufRead for SymEncryptedDataReader<R> {
     fn fill_buf(&mut self) -> io::Result<&[u8]> {
+        self.check_done()?;
```
Comment (hko-s) [inline src/composed/message/reader/sym_encrypted.rs:139]: I don't fully grasp how this state-machine and `BufRead` interact.

That said: Naively, this entire `check_done` logic seems redundant to me. I would expect that the check for `res == 0` below (followed by `set_done`) should cover all cases.

If that's right, it would seem nicer to simplify the PR.

(Experimentally, dropping the `check_done` call doesn't break the unit tests - but those might not cover all edge cases?)

File: src/composed/message/reader/sym_encrypted.rs [hunk 2]
```diff
@@ -116,7 +158,13 @@ impl<R: BufRead> BufRead for SymEncryptedDataReader<R> {
 impl<R: BufRead> Read for SymEncryptedDataReader<R> {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
         match self {
-            Self::Body { decryptor } => decryptor.read(buf),
+            Self::Body { decryptor } => {
+                let res = decryptor.read(buf)?;
+                if res == 0 {
```
[the remark this answers is not in the document]
Comment (hko-s) [reply src/composed/message/reader/sym_encrypted.rs:163]: I also don't understand why the SEIPD reader doesn't need the equivalent logic. If it indeed doesn't need it, we should maybe add a comment that justifies the difference?

**Review by hko-s — APPROVED** [review]

**Comment by hko-s** [comment]
No results while running the `1_message_from_bytes` fuzz test for >1h of CPU time.

I think we should merge this.
